Login: Install Woo from the store picker

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #7429 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR improves the store picker by including non-woo sites in the list. Tapping on a non-woo site will navigate to the No Woo error screen, where WooCommerce can be installed.

Since this PR changes all logic regarding to the table view, I added a view model for the `StorePickerViewController` so that the logic can be tested easily.

Other changes: 
- `preselectStoreIfPossible` was updated to only preselect a store if it has WooCommerce.
- Added a new property `num_of_non_woo_sites` when tracking store picker view.
- Updated `StoreTableViewCell` to show warning icons for non-woo sites.
- The store list now sorts all stores on the top and non-Woo sites at the bottom. The list sections are named based on whether the sites have Woo.
- The only store picker that doesn't show non-Woo sites is the one for switching store when logged in.

Internal discussion: p1659946510389419-slack-C03L1NF1EA3

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
Prerequisite: make sure your WP.com account have access to at least one Woo Store and at least one non-Woo sites.
**Testing WP.com login flow**
1. Log out from your account if necessary, then select Continue with WP.com.
2. Log in with your WP.com credentials.
3. Notice that you are navigated to the store picker, which shows both the stores and non-Woo sites. In Xcode console, notice both the number of stores and non-woo sites are tracked: `🔵 Tracked site_picker_stores_shown, properties: [AnyHashable("num_of_non_woo_sites"): 3, AnyHashable("num_of_stores"): 1]`
4. Select a non-woo site that is self-hosted or Atomic. Notice that a No Woo error screen is displayed, with the Install WooCommerce button present. The "See connected stores" button shouldn't be present here. Notice in Xcode console: `🔵 Tracked site_picker_non_woo_site_tapped, properties: [AnyHashable("is_non_atomic"): false]`
5. Select Install WooCommerce. When the installation completes, the No Woo screen is dismissed and the store list is updated with the updated store. This store should be selected automatically.
6. If you have a non-Atomic site, selecting it from the store picker should display the No Woo error screen without the Install WooCommerce button. Xcode console should log `🔵 Tracked site_picker_non_woo_site_tapped, properties: [AnyHashable("is_non_atomic"): true]`
7. If you have a JCP site, selecting it from the store picker should display the No Woo error screen without the Install WooCommerce button.

**Testing the Account Mismatch flow**
1. Log out from your account if necessary, then select Enter a site address.
2. Enter the address for a Woo store to which you don't have access.
3. Proceed to log in with your WP.com account.
4. Notice that the Account Mismatch screen is displayed. Select "See connected stores" button.
5. Notice that you are navigated to the store picker, which shows both the stores and non-Woo sites. Feel free to repeat steps 4, 5, 6 from the WP.com test case above.

**Testing the No Woo flow**
1. Log out from your account if necessary, then select Enter a site address.
2. Enter the address for a non-woo site that you have access to.
3. Proceed to log in with your WP.com account.
4. Notice that the Account Mismatch screen is displayed. Select "See connected stores" button.
5. Notice that you are navigated to the store picker, which shows both the stores and non-Woo sites. Feel free to repeat steps 4, 5, 6 from the WP.com test case above.

**Testing the app start flow**
1. Log out from your account if necessary, then select Continue with WP.com.
2. Log in with your WP.com credentials.
3. Notice that you are navigated to the store picker. Kill the app.
4. Open the app again, notice that a store picker is displayed with both Woo and non-Woo sites.
5.  Feel free to repeat steps 4, 5, 6 from the WP.com test case above.

**Testing the switching store flow**
1. Log in to your WP.com account and select a Woo store.
2. In the Settings screen, select the "Switch store" button.
3. Notice that the store picker displays only woo stores.

Please feel free to also test the case for a WP.com account without either Woo or non-woo sites. The empty screen should be displayed as before.


### Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://user-images.githubusercontent.com/5533851/183809469-9f22d97b-6902-48f4-a607-577da717502d.mp4

Switching store:

<img src="https://user-images.githubusercontent.com/5533851/183815498-25109997-4352-479c-86a5-08efd34e72b2.png" width=320 />

No sites:

<img src="https://user-images.githubusercontent.com/5533851/183834565-dfaff29b-ed71-4545-a3ff-c3ef6cfae25e.png" width=320 />


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

itsmeichigo committed:
Add button to install Woo from store picker

itsmeichigo committed:
Merge branch 'trunk' into issue/7429-install-woo-from-store-picker

itsmeichigo committed:
Remove install Woo button from the store picker cell

itsmeichigo committed:
Inject configuration to StorePickerViewController initializer

itsmeichigo committed:
Create view model for StorePickerViewController

itsmeichigo committed:
Update results controller for store picker based on configuration

itsmeichigo committed:
Don't display checkmark for non-woo sites

itsmeichigo committed:
Update StorePickerViewModel with configs for table view

itsmeichigo committed:
Update section titles for store picker

itsmeichigo committed:
Integrate view model to StorePickerViewController

itsmeichigo committed:
Only show non-woo sites for .listStores config

itsmeichigo committed:
Handle Woo installation from store picker

itsmeichigo committed:
Scroll to selected row after installing Woo

itsmeichigo committed:
Track install Woo from store picker

itsmeichigo committed:
Update tracks for sitePickerStoresShown

itsmeichigo committed:
Simplify No woo flow from store picker

itsmeichigo committed:
Add unit tests for StorePickerViewModel

itsmeichigo committed:
Merge branch 'trunk' into issue/7429-install-woo-from-store-picker

itsmeichigo committed:
Add more tests for StorePickerViewModel

itsmeichigo committed:
Show notice on non-woo sites

itsmeichigo committed:
Only preselect a site if it has woo

itsmeichigo committed:
Show navigation bar before pushing the No Woo screen

itsmeichigo committed:
Only show install button for site with jetpack

itsmeichigo committed:
Rename property for non-woo site number

itsmeichigo committed:
Merge branch 'trunk' into issue/7429-install-woo-from-store-picker

itsmeichigo committed:
Track selection of non-woo sites

itsmeichigo committed:
Update check for non-atomic site

itsmeichigo committed:
Drop support for installing Woo for JCP sites

jaclync commented on WooCommerce/Classes/Authentication/AuthenticationManager.swift:
```diff
@@ -197,7 +197,7 @@ class AuthenticationManager: Authentication {
                 siteURL: siteURL,
                 showsConnectedStores: matcher.hasConnectedStores,
-                showsInstallButton: matchedSite.isJetpackConnected,
+                showsInstallButton: matchedSite.isJetpackConnected && matchedSite.isJetpackThePluginInstalled,
```
does this change exclude JCP sites because these sites have `isJetpackConnected && !isJetpackThePluginInstalled`?

itsmeichigo replied:
Yes, we discovered an issue with Calypso when the Woo installation flow doesn't work properly with JCP sites (p1660123571755889-slack-C02DQP0FP) - so we decided to not support JCP sites.

jaclync replied:
I see now, do you think we can add a comment somewhere about this reason?

jaclync replied:
Since this logic is also used in `StorePickerViewController` https://github.com/woocommerce/woocommerce-ios/pull/7439#discussion_r943035935, maybe a boolean extension on `Site` would be helpful with a comment on why JCP sites aren't supported?

itsmeichigo replied:
I intended to bring this logic to `NoWooErrorViewModel` instead, so I'll make the change as part of the Woo setup enhancement in #7455. How does that sound?

jaclync commented on WooCommerce/Classes/Authentication/Epilogue/StorePickerViewController.swift:
```diff
@@ -370,27 +333,30 @@ private extension StorePickerViewController {
-        currentlySelectedSite = firstSite
+        // Otherwise select the first site in the list if it has Woo
+        if firstSite.isWooCommerceActive {
+            currentlySelectedSite = firstSite
```
how about updating `firstSite` to `sites.first { $0.isWooCommerceActive }` so that it's always non-nil if there is at least one Woo site?

itsmeichigo replied:
Good point! Updated in cdefc57f58b64f55994d8581cc2556e322335034.

jaclync commented on WooCommerce/Classes/Authentication/Epilogue/StorePickerViewController.swift:
```diff
@@ -712,6 +682,34 @@ private extension StorePickerViewController {
+        let viewModel = NoWooErrorViewModel(
+            siteURL: site.url,
+            showsConnectedStores: false, // avoid looping from store picker > no woo > store picker
+            showsInstallButton: site.isJetpackConnected && site.isJetpackThePluginInstalled,
```
similar question about excluding JCP sites here

itsmeichigo replied:
Yes we need to exclude JCP sites due to an issue with the web setup flow.

jaclync commented on WooCommerce/Classes/Authentication/Epilogue/StorePickerViewController.swift:
```diff
@@ -712,6 +682,34 @@ private extension StorePickerViewController {
+            showsInstallButton: site.isJetpackConnected && site.isJetpackThePluginInstalled,
+            onSetupCompletion: { [weak self] siteID in
+                guard let self = self else { return }
+                self.navigationController?.popViewController(animated: true)
```
non-blocking question: does it make sense to just log the user into the non-Woo store they selected & just installed Woo on? Since they already selected the site, it's probably the site they'd like to log in

itsmeichigo replied:
Great thinking! I was planning to update this as Android was doing the same. Updated in 171262a0153e52c886f3bee081fbff1c3dd01542.

jaclync commented on WooCommerce/Classes/Authentication/Epilogue/StorePickerViewController.swift:
```diff
@@ -746,87 +744,3 @@ private enum StorePickerState {
-    /// Number of TableView Sections
-    ///
-    var numberOfSections: Int {
-        return StorePickerConstants.numberOfSections
```
nit: maybe `StorePickerConstants` isn't needed anymore?

itsmeichigo replied:
We still need `estimatedRowHeight`, so I removed the unused constants in a4cd8af70b5877564fdedbbae22247b59d31ed21.

jaclync commented on WooCommerce/Classes/Authentication/Epilogue/StorePickerViewModel.swift:
```diff
@@ -0,0 +1,207 @@
+        state = StorePickerState(sites: resultsController.fetchedObjects)
+    }
+
+    func synchronizeSites(selectedSiteID: Int64?, onCompletion: @escaping (Result<Void, Error>) -> Void) {
```
non-blocking suggestion for the future: this looks like a good use case to return an `async` result

itsmeichigo replied:
Agreed! We should also remove some other logic from the controller to the view model - I didn't want to make too many unrelated changes so maybe this is something we can do in the future for improvement.

jaclync commented on WooCommerce/Classes/Authentication/Epilogue/StorePickerViewModel.swift:
```diff
@@ -0,0 +1,207 @@
+    /// Indicates if there is more than one Store.
+    ///
+    var multipleStoresAvailable: Bool {
+        return resultsController.fetchedObjects.count > 1
```
just wondering, is there a difference using `resultsController.numberOfObjects` and `resultsController.fetchedObjects.count`? asking because most of the other functions use `numberOfObjects`

itsmeichigo replied:
They should be the same - I was replacing the `count` calls with `numberOfObjects` but missed this 😅 Updated in 95c5b3b794a24df68727f14adb0e27476a11712a.

jaclync commented on WooCommerce/Classes/Authentication/Epilogue/StorePickerViewModel.swift:
```diff
@@ -0,0 +1,207 @@
+            return nil
+        }
+
+        let isWooCommerceActive = (rawStatus as NSString).boolValue
```
I'm not sure if I understand how the boolean is derived from a section name 🤔 what does the section name raw value look like and why is it a string? I'm guessing it's from the Core Data FRC API constraint to only provide a string, but maybe it's worth adding a comment for future readers

itsmeichigo replied:
Added comment in 9d740cbe506bb2f6320e01910a3a30856b582be4 😄

jaclync commented on WooCommerce/Classes/Authentication/Epilogue/StorePickerViewModel.swift:
```diff
@@ -0,0 +1,207 @@
+
+    /// Returns the site to be displayed at a given IndexPath
+    ///
+    func site(at indexPath: IndexPath) -> Yosemite.Site? {
```
nit: is `Yosemite` needed?

```suggestion
    func site(at indexPath: IndexPath) -> Site? {
```

itsmeichigo replied:
Removed in 38e1d1dcb28d2b6c8c6d4932f05f6b8798b79451.

jaclync commented on WooCommerce/Classes/Authentication/Epilogue/StoreTableViewCell.swift:
```diff
@@ -68,6 +75,13 @@ class StoreTableViewCell: UITableViewCell {
 
+    /// When enabled, the Notice ImageView will be displayed.
+    ///
+    var allowsNotice: Bool = false {
```
nit: I think `displays*` matches its functionality more? the previous `allowsCheckmark` is to "allow" all cells for padding, while `displaysCheckmark` is to actually show the checkmark. in the use case here, I think the boolean is to control whether to show the notice icon or not.

```suggestion
    var displaysNotice: Bool = false {
```

itsmeichigo replied:
Updated in d9fed176a87d86d5044c4e778061fa3918493be2.

jaclync commented on WooCommerce/WooCommerceTests/Authentication/StorePickerViewModelTests.swift:
```diff
@@ -0,0 +1,198 @@
+    }
+
+    override func tearDown() {
+        super.tearDown()
```
super nit: usually `super.tearDown()` is called the last, though it probably doesn't matter in this case

itsmeichigo replied:
Updated in 84f7e94faa1f2ea34d20cb1db382433bea63eb7a.

jaclync commented on WooCommerce/WooCommerceTests/Authentication/StorePickerViewModelTests.swift:
```diff
@@ -0,0 +1,198 @@
+import protocol Storage.StorageManagerType
+@testable import WooCommerce
+
+final class StorePickerViewModelTests: XCTestCase {
```
nice tests! 💯

jaclync commented on WooCommerce/Classes/Authentication/Epilogue/StorePickerViewController.swift:
```diff
@@ -670,19 +633,26 @@ extension StorePickerViewController: UITableViewDelegate {
         }
 
+        guard site.isWooCommerceActive else {
+            ServiceLocator.analytics.track(.sitePickerNonWooSiteTapped, withProperties: ["is_non_atomic": !site.isJetpackConnected])
```
`is_non_atomic` feels a bit confusing to me, because self-hosted sites are also not atomic 🤔 if I understand correctly, this event property is to know which stores are eligible for Woo/plugin installation like in `NoWooErrorViewModel.showsInstallButton` below? Would another name like `can_install_woo` be more clear, and an extension like `Site.isEligibleForWooInstallation` be helpful?

I'm also just curious how we determine whether a site is eligible for Woo/plugin installation. It looks like the requirement is `site.isJetpackConnected && site.isJetpackThePluginInstalled` below, does it account for user roles that cannot install plugins like shop managers? I'm sure these have been considered, I'd just like to know more since the code just checks for the site's Jetpack status here 😄

itsmeichigo replied:
The idea is to check for the number of users trying to use simple sites here (WP.com site without Business or E-commerce plans). For HEs the term simple site is more common, but I'm not sure if that's the case for us as well.

> I'm also just curious how we determine whether a site is eligible for Woo/plugin installation

This is a good idea! I'm thinking about adding this for the next improvements of the Woo installation flow. Maybe in the next PR.

itsmeichigo replied:
In fact I think we don't need to check for eligibility, since the web view handles that already:

<img src="https://user-images.githubusercontent.com/5533851/184085667-2bbada2f-bb98-4299-b85c-b0fcfa21a840.png" width=320 />

jaclync reviewed (commented):
Nice enhancement! I left some non-blocking questions, trying to understand some code changes a bit more. There was one behavior change on the store picker on the spinner state, it'd be nice to keep the same behavior as before.

> Testing WP.com login flow
> Testing the Account Mismatch flow

✅ 

> Testing the No Woo flow

✅  I think we're expected to see the error screen with Install Woo CTA instead of the Account Mismatched screen as mentioned in the PR description?

> Testing the app start flow

✅  One time it didn't show the store picker, but I couldn't reproduce it.

> Testing the switching store flow

❓  Before the PR, the spinner seems to be shown right away when refreshing the site(s). When testing the PR, it looks like the bottom CTA is disabled first, then the spinner is shown for a bit and the selected store is selected shortly

https://user-images.githubusercontent.com/1945542/184059511-69637279-ed8e-4466-9f87-8670a982299e.mov

---

❓ Design observation:

In light mode, the empty navigation bar title makes it look a bit odd on the non-Woo site screen. Does it make sense to hide the navigation bar?

<img src="https://user-images.githubusercontent.com/1945542/184061793-e8b3e0fe-04e5-4218-9841-85fa931224ef.png" width="300" />

itsmeichigo committed:
Select installed site immediately

itsmeichigo committed:
Check for woo stores before preselecting store

itsmeichigo committed:
Remove unused constants

itsmeichigo committed:
Use `numberOfObjects` for multipleStoresAvailable

itsmeichigo committed:
Add comments explaining section name

itsmeichigo committed:
Remove redundant Yosemite prefix

itsmeichigo committed:
Rename `allowsNotice` to `displaysNotice`

itsmeichigo committed:
Move `super.tearDown()` to the end

itsmeichigo committed:
Refresh result first before updating UI

itsmeichigo commented:
Thank you @jaclync for the review! I've addressed the issues mentioned in the comments, hopefully I didn't miss anything.

>  One time it didn't show the store picker, but I couldn't reproduce it.

I'm curious, what did you see instead?

> ❓ Before the PR, the spinner seems to be shown right away when refreshing the site(s). When testing the PR, it looks like the bottom CTA is disabled first, then the spinner is shown for a bit and the selected store is selected shortly

This was because I was observing state change first before syncing sites - so the UI was updated without a selected site at first. Moved this down after the synchronization in 780fd13f08e094fa00e7e208d19e1a8cceaaa4e2 to restore the old experience.

> In light mode, the empty navigation bar title makes it look a bit odd on the non-Woo site screen. Does it make sense to hide the navigation bar?

It would be nice to hide the navigation bar border, but I'm not sure if it's worth the hassle with updating the navigation bar appearance. We do need the navigation bar though, to navigate from this flow back to the store picker. I didn't want to touch the navigation bar appearance in fear of messing with other navigation controllers - but please feel free to let me know if you have a good suggestion 🙏 .

jaclync reviewed (approved):
LGTM, thanks for the updates and explanations!

>> One time it didn't show the store picker, but I couldn't reproduce it.

> I'm curious, what did you see instead?

Back to the login prologue screen, without the store picker.

> Testing the switching store flow

Working now! ✅ 

> It would be nice to hide the navigation bar border, but I'm not sure if it's worth the hassle with updating the navigation bar appearance. We do need the navigation bar though, to navigate from this flow back to the store picker. I didn't want to touch the navigation bar appearance in fear of messing with other navigation controllers - but please feel free to let me know if you have a good suggestion 🙏 .

How about showing the site name in the navigation bar title? Some text in the navigation bar might make it look more polished.

itsmeichigo commented:
> How about showing the site name in the navigation bar title? Some text in the navigation bar might make it look more polished.

Good point 👍 Added this change to #7455. Merging this PR for now.
